<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
 <head>
  <meta http-equiv="content-type" content="text/html; CHARSET=UTF-8">
  <title>向生成器中抛入一个异常</title>
<link media="all" rel="stylesheet" type="text/css" href="styles/03e73060321a0a848018724a6c83de7f-theme-base.css" />
<link media="all" rel="stylesheet" type="text/css" href="styles/03e73060321a0a848018724a6c83de7f-theme-medium.css" />

  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <link media="all" rel="stylesheet" type="text/css" href="style.css"/>
 </head>
 <body class="docs" class="docs"><div class="navbar navbar-fixed-top">
  <div class="navbar-inner clearfix">
    <ul class="nav" style="width: 100%">
      <li style="float: left;"><a href="generator.send.html">? Generator::send</a></li>
      <li style="float: right;"><a href="generator.valid.html">Generator::valid ?</a></li>
    </ul>
  </div>
</div>
<div id="breadcrumbs" class="clearfix">
  <ul class="breadcrumbs-container">
    <li><a href="index.html">PHP Manual</a></li>
    <li><a href="class.generator.html">Generator</a></li>
    <li>向生成器中抛入一个异常</li>
  </ul>
</div>
<div id="layout">
  <div id="layout-content"><div id="generator.throw" class="refentry">
 <div class="refnamediv">
  <h1 class="refname">Generator::throw</h1>
  <p class="verinfo">(PHP 5 &gt;= 5.5.0, PHP 7, PHP 8)</p><p class="refpurpose"><span class="refname">Generator::throw</span> &mdash; <span class="dc-title">向生成器中抛入一个异常</span></p>

 </div>

 <div class="refsect1 description" id="refsect1-generator.throw-description">
  <h3 class="title">说明</h3>
  <div class="methodsynopsis dc-description">
   <span class="modifier">public</span> <span class="methodname"><strong>Generator::throw</strong></span>(<span class="methodparam"><span class="type"><a href="class.throwable.html" class="type Throwable">Throwable</a></span> <code class="parameter">$exception</code></span>): <span class="type"><a href="language.types.declarations.html#language.types.declarations.mixed" class="type mixed">mixed</a></span></div>

  <p class="para rdfs-comment">
   抛出一个异常到生成器并恢复生成器的执行。
   与当前 <a href="language.generators.syntax.html#control-structures.yield" class="link">yield</a> 表达式被 <code class="literal">throw $exception</code> 语句替换是一样的行为。
  </p>
  <p class="para">
   如果调用此方法时生成器已经关闭，则将会在调用者的上下文中抛出异常。
  </p>
 </div>


 <div class="refsect1 parameters" id="refsect1-generator.throw-parameters">
  <h3 class="title">参数</h3>
  <dl>
   
    <dt>
<code class="parameter">exception</code></dt>

    <dd>

     <p class="para">
      抛出异常到生成器。
     </p>
    </dd>

   
  </dl>

 </div>


 <div class="refsect1 returnvalues" id="refsect1-generator.throw-returnvalues">
  <h3 class="title">返回值</h3>
  <p class="para">
   返回生成的值。
  </p>
 </div>


  <div class="refsect1 examples" id="refsect1-generator.throw-examples">
  <h3 class="title">示例</h3>
  <p class="para">
   <div class="example" id="example-804">
    <p><strong>Example #1 抛出异常到生成器</strong></p>
    <div class="example-contents">
<div class="phpcode"><code><span style="color: #000000">
<span style="color: #0000BB">&lt;?php<br /></span><span style="color: #007700">function </span><span style="color: #0000BB">gen</span><span style="color: #007700">() {<br />    echo </span><span style="color: #DD0000">"Foo\n"</span><span style="color: #007700">;<br />    try {<br />        yield;<br />    } catch (</span><span style="color: #0000BB">Exception $e</span><span style="color: #007700">) {<br />        echo </span><span style="color: #DD0000">"Exception: </span><span style="color: #007700">{</span><span style="color: #0000BB">$e</span><span style="color: #007700">-&gt;</span><span style="color: #0000BB">getMessage</span><span style="color: #007700">()}</span><span style="color: #DD0000">\n"</span><span style="color: #007700">;<br />    }<br />    echo </span><span style="color: #DD0000">"Bar\n"</span><span style="color: #007700">;<br />}<br /> <br /></span><span style="color: #0000BB">$gen </span><span style="color: #007700">= </span><span style="color: #0000BB">gen</span><span style="color: #007700">();<br /></span><span style="color: #0000BB">$gen</span><span style="color: #007700">-&gt;</span><span style="color: #0000BB">rewind</span><span style="color: #007700">();<br /></span><span style="color: #0000BB">$gen</span><span style="color: #007700">-&gt;</span><span style="color: #0000BB">throw</span><span style="color: #007700">(new </span><span style="color: #0000BB">Exception</span><span style="color: #007700">(</span><span style="color: #DD0000">'Test'</span><span style="color: #007700">));<br /></span><span style="color: #0000BB">?&gt;</span>
</span>
</code></div>
    </div>

    <div class="example-contents"><p>以上示例会输出：</p></div>
    <div class="example-contents screen">
<div class="cdata"><pre>
Foo
Exception: Test
Bar
</pre></div>
    </div>
   </div>
  </p>
 </div>

</div><div id="usernotes">
 <div class="head">
  <h3 class="title">User Contributed Notes</h3>
 </div>
<div id="allnotes">  <a name="121864""></a>
  <div class="note">
   <strong class="user">gt199899 at gmail dot com</strong>
   <a href="#121864" class="date">13-Nov-2017 11:38</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
$gen = (function () {<br />
&nbsp;&nbsp;&nbsp; try {<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; yield 1;<br />
&nbsp;&nbsp;&nbsp; } catch (Exception $e) {<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; echo $e-&gt;getMessage();<br />
&nbsp;&nbsp;&nbsp; }<br />
})();<br />
<br />
$gen-&gt;throw(new Exception('gen throw exception'));</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
  <a name="121863""></a>
  <div class="note">
   <strong class="user">gt199899 at gmail dot com</strong>
   <a href="#121863" class="date">13-Nov-2017 11:37</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
$gen = (function () {<br />
&nbsp;&nbsp;&nbsp; try {<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; yield 1;<br />
&nbsp;&nbsp;&nbsp; } catch (Exception $e) {<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; echo $e-&gt;getMessage();<br />
&nbsp;&nbsp;&nbsp; }<br />
})();<br />
<br />
$gen-&gt;throw(new Exception('gen throw exception'));</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
  <a name="119805""></a>
  <div class="note">
   <strong class="user">jeudipolanco at ymail dot com</strong>
   <a href="#119805" class="date">30-Aug-2016 09:31</a>
   <div class="text">
    <div class="phpcode">
<code><span class="html">
there have errors the form correct to do that is of this way:<br />
<br />
&nbsp;function gen() {<br />
&nbsp;&nbsp;&nbsp; echo "Foo\n";<br />
&nbsp;&nbsp;&nbsp; try {<br />
&nbsp;&nbsp; &nbsp; <br />
&nbsp;&nbsp;&nbsp; throw new Exception('Prueba');<br />
<br />
&nbsp;&nbsp;&nbsp; } catch (Exception $e) {<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; echo "Excepción: {$e-&gt;getMessage()}\n";<br />
&nbsp;&nbsp;&nbsp; }<br />
&nbsp;&nbsp;&nbsp; echo "Bar\n";<br />
}<br />
<br />
&nbsp;gen();<br />
<br />
&nbsp;i resolved one tecnic of how work getMessage look at it:<br />
<br />
$myarray=array();<br />
<br />
if($respuesta == "Mark"){<br />
<br />
for($i=5;$i&lt;=16;$i++){<br />
try {<br />
<br />
&nbsp;&nbsp;&nbsp; $palabra="soy la excepcion en la linea:".$i;<br />
<br />
&nbsp;&nbsp; &nbsp; throw new Exception($palabra);<br />
<br />
}<br />
catch(Exception $e) {<br />
&nbsp;<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; $myarray[$e-&gt;getLine()] =$palabra;<br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; <br />
&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; if($e-&gt;getLine() == $i){<br />
&nbsp; echo "&lt;br&gt;&lt;b&gt;La excepción se creó en la línea : " .$e-&gt;getLine()."LA LINEA DICE:". $myarray[$e-&gt;getLine()]."&lt;/b&gt;";<br />
}//if<br />
<br />
}//catch<br />
&nbsp;&nbsp;&nbsp; <br />
<br />
}//for<br />
<br />
}//if</span>
</code>
    </div>
   </div>
  </div>
 <div class="foot"></div>
</div>
</div>
</div></div></body></html>
